### ISSUE TITLE
Try using DiskANN

### ISSUE THREAD

**Issue by geohacker** [developmentseed/bioacoustics-api#55]
Milvus supports on-disk index https://milvus.io/docs/disk_index.md. This should reduce compute usage drastically.  DiskANN is enabled by default but I think we need to make sure the volume inside the container is mounted at `/var/lib/milvus/data`

I think we'll want to add a [non-boot persistent disk](https://cloud.google.com/compute/docs/disks/persistent-disks) to the cluster and then have k8s mount that to the query nodes and index nodes. We'll need to use `DISKANN` as the index type while building the index.

@leothomas Once you have the index update done #54, let's figure out a plan here. Might be good to create a second indexing script/job that can create the DiskANN index that has a volume mounted.

I should be able to quickly adjust the k8s configuration to support this and get some help @sunu too.

**Comment by sunu** [comment]
@geohacker from a quick look, it doesn't look like the Milvus Helm chart supports mounting a persistent disk to `/var/lib/milvus/data` like you mentioned. It's [mounting an `emptyDir`](https://github.com/zilliztech/milvus-helm/blob/1b1d28ed269b20c1084e758832342a6ac139e7db/charts/milvus/templates/indexnode-deployment.yaml#L179-L182) there.

Is it possible that the volume is only used to build a local cache of the index and the actual index is stored somewhere else (eg on Minio blob storage for example)?

**Comment by geohacker** [comment]
@sunu Oh I'm not sure if I'm reading this right but this looks exactly like what we want https://github.com/zilliztech/milvus-helm/blob/master/charts/milvus/templates/indexnode-deployment.yaml#L137-L139. What do you think?

**Comment by sunu** [comment]
@geohacker yes, we do get a volume at the right path but the volume is an ephemeral volume on the boot disk since it uses `emptyDir`. I talked a bit about this with @leothomas yesterday and we decided we will give it a try and see if we run into any issues. Leo will also ask for clarification about this on the Milvus discord. Our guess is that Milvus stores the actual index on Minio blob storage and the on-disk index volume on index nodes and query nodes are only used for maintaining a local cache of that index. So it's fine for that volume to be ephemeral.

I have deployed the stack back with [`enableDisk` set to `true`](https://github.com/developmentseed/bioacoustics-deploy/commit/8b5bcc369672b8bb17465ce8ca159d01df7702d0) without tweaking any other resource limits etc.

**Comment by leothomas** [comment]
Alright! We've run a successful test with the DiskANN index, and here's the TL;DR: while it works, it doesn't necessary offer the advantages we were hoping for. Namely Milvus's DiskANN implementation contains an un-avoidable product quantization step. @sdenton4 my understanding was that you wanted to avoid any additional compression on the embeddings, which would the make the DiskANN a no-go. Should we continue exploring this indexing strategy? Do you prefer to explore having the embeddings model generate embeddings with fewer dimensions? (And possibly generate integer  [0:255] values, rather than float, which reduces the byte count by a factor of 4)

The long form: 

I started with noticing that the switch to the DiskANN was making the kubernetes index node crash and restart due to hitting CPU limits: 
![image](https://github.com/developmentseed/bioacoustics-api/assets/11858457/9984f6dc-021a-4693-a26e-9ab853cc4ee5)

Bumping that limit got us past the crashing and up and running with DiskANN index successfully. I didn't get a chance to test too extensively, but the queries seemed to execute comparably quickly (difficult to compare since I ingested a subset of the data, to verify that the DiskANN index was correctly configured): 
![image](https://github.com/developmentseed/bioacoustics-api/assets/11858457/08dc681f-3f41-4243-bf3c-46bc06a40124)

(Notice the disk memory usage increasing up from 0 to ~4Gb once the data it loaded into the DiskANN index)

However when I checked a couple days later, I found the index node crashing and restarting again. 
![image](https://github.com/developmentseed/bioacoustics-api/assets/11858457/c490f5f9-17f0-4f68-919d-6bef30fb6cc8)

I think this was due to the fact that increasing the CPU _limit_ (to 10) but not the _requested_ CPU gave us some extra burst capacity to use, and when that burst capacity ran out the node started crashing whenever we hit the _requested_ CPU (5).

Looking deeper into the node logs as to why we were using some much CPU when the index was starting up, I noticed this line in the logs: 
![image](https://github.com/developmentseed/bioacoustics-api/assets/11858457/fe409c71-d9b6-4ff8-9e65-93af6b8ae19b)

Which seems to indicate the fact that the DiskANN index requires all that CPU to perform a Product Quantization operation on the 512 dimension embeddings (compressing them from 16384 bytes/vector (512 dims * 32 bytes/float) to 256 bytes/vector). 

Since the whole point of using DiskANN was to avoid having to compress or otherwise reduce the embeddings, I posted in the Milvus Discord channel to see if this could somehow be avoided. The response was essentially no, that's how Milvus impelments a DiskANN index. Response: 

> DISKINDEX is a combine of PQ and Vamana graph index. PQ index is loaded into memory and Vamana graph on disk. When you call search(), it first searches in PQ, then finds some candidates to search again in the Vamana graph on disk. Product quantization step cannot be avoided. If you read the doc for DISKINDEX, you will see some parameters indicate it uses product quantization:
```
PQCodeBugetGBRatio ---------------- Size limit on the PQ code. A larger value offers a higher recall rate but increases memory usage.
```
> https://milvus.io/docs/disk_index.md#On-disk-Index

**Comment by sdenton4** [comment]
Product quantization makes sense; I think it's the _PCA_ that we want to avoid. So this still seems OK to me!

**Comment by leothomas** [comment]
Hey there @sdenton4! Got an update for you: 

## TL;DR:  
### The good:
We have ~3.5 million records, from 30 of the 512 embedding files (out of ~3.7k total files) loaded into Milvus using the DiskANN index. The search is performant, but not quite as performant as the current IVF index (~0.1 seconds for ~100 results and  ~7 seconds for 1000 results and 250 seconds for the max result size of 16384). I haven't verified the quality of the index result, since there's no way to do a flat search, but I'm confident the implementation is sound, based on the performance of the other indexes. (I can put together a notebook if we want to evaluate the _quality_ of the DiskANN index results, if needed)
### The bad:
In order to search on a collection it has to be [loaded into memory](https://milvus.io/docs/load_collection.md), which is proving to be incredibly slow, past ~1.5 million records. With the current ~3.5 million records we're looking at almost 1 hour load time!

I've started a thread in Discord to try to understand why it takes so long to load the collection, especially since the whole point of the on-disk index is that it shouldn't have to be loaded into memory.

There are no errors in the logs, however one clue (that I've shared in the Milvus discord channel) that might point to a bug in Milvus is:
```
[2024/02/15 20:32:11.583 +00:00] [INFO] [querynode/segment_loader.go:301] ["loading delta..."] [segmentID=447713668967530627]
[2024/02/15 20:32:11.583 +00:00] [INFO] [querynode/segment_loader.go:526] ["there are no delta logs saved with segment, skip loading delete record"] [segmentID=447713668967530627]
```
Which occurs when loading each segment, and seems to imply that the segment is not sealed. (Context: when loading data into Milvus, the data gets partitioned into segments, with a pre-determined max size. When the segment reaches this size, it gets "sealed" and indexed. This way new data can be searchable as soon as the segment is sealed, even if more data is still being loaded).

Logically, if the segments aren't being sealed, then they're not being indexed, which could be why they're being loaded into memory (rather than searched from disk).

The Milvus devs followed up in Discord, with a rather unhelpful/generic message outlining the memory requirements, rather than addressing the question of why the index segments have to be loaded into memory if were using an **on-disk** index. They did mention that the unsealed segments might be segments that are simply still in the process of indexing (after the ingestion). I tried releasing the collection and re-loading in this morning, and, while the logs messages indicating the segments hadn't been sealed are no longer showing up in the logs, the collection is still takes >1hr to load. 

Going forward, we have a few different options: 

- We can continue to try work with the DiskANN index, exploring different configurations, memory allocations, partitioning schemes, etc. 
- We can pivot back to an HNSW or IVF index for performance, and look into an index which includes product quantization in order to reduce the memory requirements of the database (@sdenton4 you had mentioned that production quantization is not a problem, since, unlike PCA, it doesn't drop any dimensions - is that correct?)
- We can leave the backend as is (loaded with 1% data sample and DiskANN index) and update the API to remove the PCA transform of input/search vectors, and revisit later. 
- We can look into pivoting to a different vector database options - we've had some promising results with `pgvector` at a similar scale (250million records), but there is a _huge_ proliferation of other options we can look at: Pinecone, Weviate, Astra, Vespa, Qdrant, Redis, ElasticsSearch, Cassandra, MyScaleDB, LanceDB, Marqo, etc)

(Note on partitioning: partitioning essentially creates sub-datasets that we can ingest data into and search separately. This is useful for data which is often filtered at search time such as searching for either a TV show _or_ a movie on Netflix. In this case we would create two partitions, and only search one or the other. Datasets where users often want to search across filters, such as literature genres, would not be a good candidate) 

Happy to hop on a call to discuss this further next week @sdenton4 !

**Comment by leothomas** [comment]
Happy to hop on call to chat about this further next week @sdenton4

**Comment by sdenton4** [comment]
Thanks for looking into this, Leo!

- Yes, product quantization should be fine - it's the PCA we would like to factor out with the 512-dim'l embeddings.

- The Milvus+DiskANN does sound like a pain - sounds a bit like they built their system with an in-memory assumption and are behaving strangely when confronted with a new world... I'll try looking around a bit and seeing what else is out there. 

- The main reason we are getting 16k results right now is so that we can filter down, eg by geography. If we had a system which included metadata in the search itself, a shorter result list would be just fine.
